chain/index/events.go

```diff
@@ -0,0 +1,560 @@
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// if we've already indexed events for this tipset, mark them as unreverted and return
+	res, err := tx.Stmt(si.stmts.updateEventsToNonRevertedStmt).ExecContext(ctx, msgTsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("failed to unrevert events for tipset: %w", err)
+	}
+	rows, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("failed to get rows affected by unreverting events for tipset: %w", err)
+	}
+	if rows > 0 {
+		log.Infof("unreverted %d events for tipset: %s", rows, msgTs.Key())
+		return nil
+	}
+
+	if !si.cs.IsStoringEvents() {
+		return nil
+	}
+
+	ems, err := si.executedMessagesLoaderFunc(ctx, si.cs, msgTs, executionTs)
+	if err != nil {
+		return xerrors.Errorf("failed to load executed messages: %w", err)
+	}
+	eventCount := 0
+	addressLookups := make(map[abi.ActorID]address.Address)
+
+	for _, em := range ems {
+		msgCidBytes := em.msg.Cid().Bytes()
+
+		// read message id for this message cid and tipset key cid
+		var messageID int64
+		if err := tx.Stmt(si.stmts.getMsgIdForMsgCidAndTipsetStmt).QueryRowContext(ctx, msgTsKeyCidBytes, msgCidBytes).Scan(&messageID); err != nil {
+			return xerrors.Errorf("failed to get message id for message cid and tipset key cid: %w", err)
+		}
+		if messageID == 0 {
+			return xerrors.Errorf("message id not found for message cid %s and tipset key cid %s", em.msg.Cid(), msgTs.Key())
+		}
+
+		// Insert events for this message
+		for _, event := range em.evs {
+			addr, found := addressLookups[event.Emitter]
+			if !found {
+				var ok bool
+				addr, ok = si.actorToDelegatedAddresFunc(ctx, event.Emitter, executionTs)
+				if !ok {
+					// not an address we will be able to match against
+					continue
+				}
+				addressLookups[event.Emitter] = addr
+			}
+
+			var robustAddrbytes []byte
+			if addr.Protocol() == address.Delegated {
+				robustAddrbytes = addr.Bytes()
+			}
+
+			// Insert event into events table
+			eventResult, err := tx.Stmt(si.stmts.insertEventStmt).ExecContext(ctx, messageID, eventCount, uint64(event.Emitter), robustAddrbytes, 0)
+			if err != nil {
+				return xerrors.Errorf("failed to insert event: %w", err)
+			}
+
+			// Get the event_id of the inserted event
+			eventID, err := eventResult.LastInsertId()
+			if err != nil {
+				return xerrors.Errorf("failed to get last insert id for event: %w", err)
+			}
+
+			// Insert event entries
+			for _, entry := range event.Entries {
+				_, err := tx.Stmt(si.stmts.insertEventEntryStmt).ExecContext(ctx,
+					eventID,
+					isIndexedFlag(entry.Flags),
+					[]byte{entry.Flags},
+					entry.Key,
+					entry.Codec,
+					entry.Value,
+				)
+				if err != nil {
+					return xerrors.Errorf("failed to insert event entry: %w", err)
+				}
+			}
+			eventCount++
+		}
+	}
+
+	return nil
+}
+
+func MakeLoadExecutedMessages(recomputeTipSetStateFunc recomputeTipSetStateFunc) func(ctx context.Context,
```

Comment:
maybe we should ditch the indirection here and just make the indexer take a `recomputeTipSetStateFunc` directly and then run `loadExecutedMessages` on that when we need to?

Reply:
@rvagg This makes unit testing easy which is the primary reason I added this. Please see tests in `event_test.go`.

Reply:
I couldn't find this comment thread when I commented below in node/modules/chainindex.go, but here it is!

Here's a version of it which avoids this problem for your consideration: https://github.com/filecoin-project/lotus/pull/12604, this doesn't leak anything about "executed messages" outside of the package, all it wants is state-recompute, which is all it actually needs, and even that's not strictly required.

I've gone with the ~idiomatic `NewXWithFoo()` type constructor, but you could equally make a private `setExecutedMessagesLoaderFunc` that only your tests can use, or just let them set the field directly, since your tests are all run within the `indexer` package and have access to mess with that.

Reply:
@rvagg Have fixed this. Please take a look.